[ZIP 231] Updates related to the Defuse Security audit. (#1192)
zcash/zips: zips/zip-0231.md

Reviewer: What does "fully transfer a note" mean? What is the distinction between a transaction that does that and an "ordinary transaction with change"? I don't know how "fully transfer" or "ordinary" are defined here.

failure to authenticate this address can enable phishing attacks. These
Authenticated Reply Addresses require zero-knowledge proofs, and for the
Orchard protocol these proofs are too large to fit into the existing 512-byte

> memo field (they are approximately 4 kilobytes in size). To support the

Author: Do we actually want to authenticate *every* sender? That doesn't really fit with the notion of a "reply-to" address; if multiple senders contribute notes to a transaction, they may not also all want to contribute addresses that can receive replies; in fact, I can think of some use cases where a third party is specifically included as a sender just so that they can receive reply messages. This isn't relevant to this ZIP per se, but to how we define that authentication protocol.

bandwidth constraints, to be able to transmit encrypted notes to the client
without including the encrypted memo data. In the current light client protocol
[^zip-0307], this is done by truncating the note ciphertext to just the part
that encrypts the note plaintext, removing the 512-byte encrypted memo and the
16-byte AEAD authentication tag. Because the authentication tag is not
transmitted to the light client, the decryption performed by the client does
not meet the standard IND-CCA2 security notion for an authenticated encryption

> scheme.

Reviewer: ```suggestion
not meet the standard IND-CCA2 ∧ INT-CTXT security notion for an authenticated
encryption scheme [^BN2000].
```

After the activation of this ZIP, recipient count and memo count are decoupled.
It becomes possible to construct transactions for which there are many more
recipients than distinct memos, or vice versa. Without mitigation, this could
result in more observably distinct patterns relating the number of recipients

> to the number of possible memos. In particular, a transaction with no memo

Reviewer: ```suggestion
to the number of possible memos. In particular, a v6-onward transaction with no memo
```

The first byte $\mathtt{0xE0}$ should be added to the documentation of inputs to
$\mathsf{PRF^{expand}}$ in § 4.1.2 ‘Pseudo Random Functions’ [^protocol-abstractprfs].

If the generated key $\mathsf{encryption\_key}$ is 32 $\mathtt{0xFF}$ bytes,
the transaction constructor MAY repeat this procedure with a different salt, in
order to avoid the recipient misinterpreting the output as having no memo data.

> Since that has negligible probability, it alternatively MAY omit this check.

Reviewer: This MAY is inconsistent with the SHOULD on line 230-232.

> Since that has negligible probability, it alternatively MAY omit this check.

Reviewer 2: ```suggestion
Since that has negligible probability, it alternatively could omit this check.
```

Each memo MUST have a length that is a multiple of 256 bytes. (For UTF-8 text
memos, padding is performed at the application level as specified in ZIP 302
[^zip-0302].) The memo is split into 256-byte plaintext chunks. Each plaintext
chunk is encrypted with ChaCha20Poly1305 [^rfc-8439] as follows:

> $\hspace{2em}\mathsf{IETF\_AEAD\_CHACHA20\_POLY1305}(\mathsf{encryption\_key}, \mathsf{nonce}, \mathsf{memo\_chunk})$

Reviewer 2: The current chunked memo design is a pretty straightforward application of STREAM. https://github.com/C2SP/C2SP/pull/225 is an alternative chunking design (similar to STREAM and FLOE); we should look at it to see if there are any cryptographic adjustments it makes that are worth adopting here,

Finally, the encrypted memo chunks for all memos are combined into a single
sequence using an order-preserving shuffle. Memo chunks from different memos MAY
be interleaved in any order, but memo chunks from the same memo MUST have the
same relative order.

The shuffle algorithm SHOULD select each successive chunk by choosing among the

> remaining lists with probability weighted by remaining list length. That is,

Reviewer 2: ```suggestion
Given a set of memos, each broken into a sequence of chunks, the shuffle
algorithm SHOULD select each successive chunk by choosing among the sequences
with probability weighted by each sequence's remaining length. That is,
```

    (memo_a, 2),
]
```

### Memo bundle padding

If the transaction contains any shielded outputs, the memo bundle MUST contain
an even number of memo chunks, with a minimum of 2. If the number of memo chunks
resulting from the encryption and shuffle steps above does not satisfy this
requirement, the transaction builder MUST append 272-byte padding chunks
generated from a CSPRNG until the total is even and at least 2. Padding chunks
are indistinguishable from encrypted memo chunks to an observer who does not

> hold the corresponding memo key.

Reviewer 2: This holds, but there's been confusion in the past as to why due to RFC 7539 section 2.7, which says:
> Additionally, unlike HMAC, Poly1305 is biased, so using it for key derivation would reduce the security of the symmetric encryption.

This would suggest that authentication tags are not indistinguishable from random.

However, as [Samuel Neves pointed out on the noise ML](https://moderncrypto.org/mail-archive/noise/2018/001461.html) when this came up there in 2018, in this AEAD the Poly1305 keys are the first 256 bits of ChaCha20 stream output (which *is* indistinguishable from random), and the tag is `(polyeval(ciphertext, clamp(r)) + s) mod 2^128`. Thus `s` acts as blinding factor on the tag.

This rationale obviously doesn't apply if we instead still use ciphertexts for padding chunks.

    (memo_a, 2),
]
```

### Memo bundle padding

If the transaction contains any shielded outputs, the memo bundle MUST contain
an even number of memo chunks, with a minimum of 2. If the number of memo
chunks resulting from the encryption and shuffle steps above does not satisfy
this requirement, the transaction builder MUST append padding chunks generated
by encrypting random memo data using a random memo key generated using a CSPRNG
until the total is even and at least 2. Padding chunks generated in this
fashion will be indistinguishable from real encrypted memo chunks to an observer

> who does not hold the memo key for those chunks.

Reviewer: If there was previously only one memo, it will be more likely to be in the first chunk if the padding chunks are only appended (and conversely, if we are a recipient and our memo is the first chunk, it is more likely that there was padding). The padding should be added before shuffling, to avoid that probabilistic leakage to the recipient.

## Memo decryption

When a recipient decrypts a shielded output, they obtain a memo key $\mathsf{K^{memo}}$.
If $\mathsf{K^{memo}}$ consists of 32 $\mathtt{0xFF}$ bytes, no memo is associated with
this output and memo decryption SHOULD NOT be performed. If the memo bundle has

> been pruned (i.e. $\mathtt{fAllPruned} = 1$), memo decryption SHOULD NOT be

Reviewer 2: ```suggestion
been pruned (i.e. $\mathtt{fAllPruned} = 1$), memo decryption cannot be
```

attempted, as no memo chunk data is available. Otherwise, the recipient derives
`encryption_key` as above, and then proceeds as follows:

> $\hspace{1.5em}$ let mutable $\mathsf{memo} \;{\small ⦂}\; \mathbb{B}^* \leftarrow \varepsilon \\$

Reviewer 2: ```suggestion
$\hspace{1.5em}$ let mutable $\mathsf{memo} \;{\small ⦂}\; \mathbb{B}^* \leftarrow [] \\$
```

$\hspace{1.5em}$ let mutable $\mathsf{counter} \;{\small ⦂}\; \mathbb{N} \leftarrow 0 \\$
$\hspace{1.5em}$ let mutable $\mathsf{potential\_last\_chunk\_index} \;{\small ⦂}\; \mathbb{N} \leftarrow 0 \\$
$\hspace{1.5em}$ for $i$ from $0$ up to $\mathtt{nMemoChunks}-1$: $\\$
$\hspace{3.0em}$     let $\mathsf{nonce} = \mathsf{I2BEOSP}_{88}(\mathsf{counter}) \,||\, [\mathtt{0x00}] \\$
$\hspace{3.0em}$     let $P = \mathsf{IETF\_AEAD\_CHACHA20\_POLY1305.Decrypt}(\mathsf{encryption\_key}, \mathsf{nonce}, \mathtt{vMemoChunks}[i]) \\$
$\hspace{3.0em}$     if $P \neq \bot$: $\\$
$\hspace{4.5em}$         set $\mathsf{memo} \leftarrow \mathsf{memo} \,||\, P \\$
$\hspace{4.5em}$         set $\mathsf{counter} \leftarrow \mathsf{counter} + 1 \\$
$\hspace{4.5em}$         set $\mathsf{potential\_last\_chunk\_index} \leftarrow i + 1 \\$
$\,\\$
$\hspace{1.5em}$ let $\mathsf{final\_nonce} = \mathsf{I2BEOSP}_{88}(\mathsf{counter}) \,||\, [\mathtt{0x01}] \\$
$\hspace{1.5em}$ let mutable $\mathsf{success} \leftarrow \kern0.05em$ false $\\$

> $\hspace{1.5em}$ for $i$ from $0$ up to $\mathtt{nMemoChunks}-1$: $\\$

Reviewer 2: Currently the algorithm is written in a half-constant-time way. We should either have this iterate explicitly from `potential_last_chunk_index` instead of 0, or make it fully constant time if we think that would be useful.

a node's wallet recognized a particular output.

Deferring memo bundle decryption decorrelates the observable timing of block
processing from whether any notes were successfully decrypted, making it more
difficult for an adversary to attribute memo decryption work to a specific
transaction.
</details>
<br/>

## Encoding in transactions

| Bytes    | Name                   | Data Type                                                 | Description                                                              |
|----------|------------------------|-----------------------------------------------------------|--------------------------------------------------------------------------|
|   1      | $\mathtt{fAllPruned}$  | $\mathtt{uint8}$                                  | 1 if the memo bundle has been pruned, otherwise 0.                  |
|   32     | $\mathtt{saltOrHash}$  | $\mathtt{byte}[32]$                               | The salt for deriving memo encryption keys, or the memo bundle digest. |
| † varies | $\mathtt{nMemoChunks}$ | $\mathtt{compactSize}$                            | The number of memo chunks.                                          |

> | † varies | $\mathtt{vMemoChunks}$ | $\mathtt{byte}[272][\mathtt{nMemoChunks}]$        | A sequence of encrypted memo chunks.                                |

Reviewer: Make sure this renders correctly; I'm not sure whether alignment of the columns is required.

> | † varies | $\mathtt{vMemoChunks}$ | $\mathtt{byte}[272][\mathtt{nMemoChunks}]$        | A sequence of encrypted memo chunks.                                   |

Reviewer 2: If we do have this type still defined then keep its reference:
```suggestion
| † varies | $\mathtt{vMemoChunks}$ | $\mathtt{MemoChunk}[\mathtt{nMemoChunks}]$        | A sequence of encrypted memo chunks.                                   |
```

## Network protocol

Nodes MUST reject `GetData` responses having an $\mathtt{fAllPruned}$ value that is

> nonzero. A node that has pruned the memo bundle from a transaction MUST NOT serve

Reviewer 2: ```suggestion
Nodes MUST reject `tx` messages having an $\mathtt{fAllPruned}$ value that is
nonzero. A node that has pruned the memo bundle from a transaction MUST NOT serve
```

> that transaction in response to a `GetData` request; it SHOULD respond as though

Reviewer 2: ```suggestion
that transaction in response to a `getdata` message for `MSG_TX` or `MSG_WTX`; it SHOULD respond as though
```

the transaction is not available.

Because memo data MAY be pruned from stored blocks, nodes that have pruned memo
data are not required to serve complete block data for affected transactions.
However, to support reliable memo delivery, it would be beneficial for the Zcash
community to establish expectations for how long after block confirmation a node
is expected to retain complete (unpruned) memo data. Such a retention policy
would allow wallet implementations to make reasonable assumptions about memo
availability when recovering from backup or performing initial synchronization.

> 

Reviewer 2: Move this down to an "Open Questions" section.

In § 4.20.1 ‘Encryption (Sapling and Orchard)’ [^protocol-saplingandorchardinband]:

* For v6-onward note ciphertexts, the KDF used to derive the note encryption key

>   MUST use a different BLAKE2b-256 personalization string than for pre-v6 note

Reviewer 2: ```suggestion
  uses a different BLAKE2b-256 personalization string than for pre-v6 note
```

  ciphertexts, to provide domain separation. The personalization string is formed
  by concatenating a protocol prefix with the 4-byte little-endian encoding of
  the transaction version group id ($\mathsf{nVersionGroupId}$):

    > For v6-onward Sapling note ciphertexts:
    >
    > $\hspace{2em}\mathsf{KDF^{Sapling}}(\mathsf{sharedSecret}, \mathsf{ephemeralKey}) := \mathsf{BLAKE2b\text{-}256}(\text{"Zc_SaplingKD"} \,||\, \mathsf{I2LEOSP}_{32}(\mathsf{nVersionGroupId}), \mathsf{sharedSecret} \,||\, \mathsf{ephemeralKey})$
    >
    > For v6-onward Orchard note ciphertexts:
    >
    > $\hspace{2em}\mathsf{KDF^{Orchard}}(\mathsf{sharedSecret}, \mathsf{ephemeralKey}) := \mathsf{BLAKE2b\text{-}256}(\text{"Zc_OrchardKD"} \,||\, \mathsf{I2LEOSP}_{32}(\mathsf{nVersionGroupId}), \mathsf{sharedSecret} \,||\, \mathsf{ephemeralKey})$

    This prevents a malicious lightwalletd server from presenting a v6 note ciphertext
    to a wallet as though it were a partial v5 ciphertext, which could otherwise cause

>     the wallet to decrypt and act on unauthenticated plaintext.

Reviewer: Note that this is another way in which we assumed that memo bundles would be deployed simultaneously with v6. This approach cannot be gated on the note plaintext lead byte; it affects decryption before we have seen that byte, and so must be gated on the tx version (or some other explicit unencrypted marker).

TBD

# Alternatives

- An alternative chunk size could be selected. The 256-byte chunk size was
  chosen because the 16-byte AEAD tag for each chunk introduces a 6.25%
  per-chunk overhead; smaller chunk sizes would require a higher percentage
  of the memo bundle data to be devoted to encryption overhead, while
  larger chunk sizes would tend to bloat the chain.
- The 64-chunk limit could be modified to be either smaller or larger. However,
  as the smallest Orchard authenticated reply-to address proofs are
  approximately 4 kilobytes in size, the limit needs to be sufficient to
  permit this use case.
- The fee algorithm could be modified to be a superlinear function of the
  number of memo chunks, in order to economically discourage the use of
  large amounts of memo data.
- An alternative approach to memo pruning that permitted pruning at the
  per-chunk level was proposed, but was judged to provide insufficient benefit
  relative to its complexity cost. Per-chunk pruning would have allowed
  selective removal of individual memos, but this added complexity to the

>   encoding and also introduced the risk that a maliciously-pruned memo could

Reviewer 2: ```suggestion
  encoding and also introduced the risk that a maliciously pruned memo could
```

TBD

# Alternatives

- An alternative chunk size could be selected. The 256-byte chunk size was
  chosen because the 16-byte AEAD tag for each chunk introduces a 6.25%
  per-chunk overhead; smaller chunk sizes would require a higher percentage
  of the memo bundle data to be devoted to encryption overhead, while
  larger chunk sizes would tend to bloat the chain.
- The 64-chunk limit could be modified to be either smaller or larger. However,

>   as Orchard authenticated reply-to address proofs are approximately 4

Reviewer 2: ```suggestion
  as the smallest Orchard authenticated reply-to address proofs are approximately 4
```
